xls/scheduling/pipeline_schedule.cc

```diff
@@ -141,6 +144,175 @@ absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegisters(
+}
+
+// Compute Critical Combinational Path's delay between any pair of nodes.
+//
+// Complexity: O( |V|·(|V| + |E|) )
+//
+// The result maps path's (src, dst) pair to the path's combinational delay.
+std::vector<std::pair<std::pair<Node*, Node*>, int64_t>>
+ComputeCriticalCombinationalPaths(
+    FunctionBase* f, const absl::flat_hash_map<Node*, int64_t>& node_delay) {
+  using PathTy =
+      std::pair<Node*, Node*>;  // describe a path, from src-node to dst-node
+  std::vector<std::pair<PathTy, int64_t>>
+      result;  // critical combinational path's delay
+
+  for (Node* src : f->nodes()) {
+    for (auto [dst, ccp] : ComputeSingleSourceCCP(src, node_delay)) {
+      std::pair<Node*, Node*> path(src, dst);
+      result.push_back(std::make_pair(path, ccp));
+    }
+  }
+  return result;
+}
+
+using namespace operations_research;
+
+// Schedule to minimize the total pipeline registers using SDC scheduling
+// the constraint matrix is totally unimodular, this ILP problem can be solved
+// by LP.
+// Reference:
+//   - Cong, Jason, and Zhiru Zhang. "An efficient and versatile scheduling
+//   algorithm based on SDC formulation." 2006 43rd ACM/IEEE Design Automation
+//   Conference. IEEE, 2006.
+//   - Zhang, Zhiru, and Bin Liu. "SDC-based modulo scheduling for pipeline
+//   synthesis." 2013 IEEE/ACM International Conference on Computer-Aided Design
+//   (ICCAD). IEEE, 2013.
+absl::StatusOr<ScheduleCycleMap> ScheduleToMinimizeRegistersSDC(
+    FunctionBase* f, int64_t pipeline_stages,
+    const DelayEstimator& delay_estimator, sched::ScheduleBounds* bounds,
+    int64_t clock_period_ps) {
+  XLS_VLOG(3) << "ScheduleToMinimizeRegistersSDC()";
+  XLS_VLOG(3) << "  pipeline stages = " << pipeline_stages;
+  XLS_VLOG_LINES(4, f->DumpIr());
+
+  XLS_VLOG(4) << "Initial bounds:";
+  XLS_VLOG_LINES(4, bounds->ToString());
+
+  std::unique_ptr<MPSolver> solver(MPSolver::CreateSolver("SCIP"));
+  if (!solver) {
+    XLS_LOG(WARNING) << "SCIP solver unavailable.";
+    return absl::UnavailableError("SCIP solver unavailable.");
+  }
+
+  const double infinity = solver->infinity();
+
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_cycle_var;  // Node's cycle after scheduling
+  absl::flat_hash_map<Node*, MPVariable*>
+      node_lifetime_var;  // Node's lifetime, since it finishes until it's
+                          // consumed by the last user
+  for (Node* node : f->nodes()) {
+    node_cycle_var[node] =
+        solver->MakeNumVar(bounds->lb(node), bounds->ub(node), node->GetName());
+    node_lifetime_var[node] = solver->MakeNumVar(
+        0.0, infinity, absl::StrFormat("lifetime_%s", node->GetName()));
+  }
+
+  for (Node* node : f->nodes()) {
+    MPVariable* const var_node = node_cycle_var.at(node);
+    MPVariable* const var_node_lifetime = node_lifetime_var.at(node);
+    for (Node* user : node->users()) {
+      MPVariable* const var_user = node_cycle_var.at(user);
+
+      // Constraint: cycle[i_user] - cycle[i] <= lifetime[i]
```

Comment:
You also need a constraint `cycle[i_user] - cycle[i] >= 0`, otherwise causality can be violated. Note that this is not equivalent to the `lifetime[i] >= 0` constraint present on line 259.